### PR TITLE
 [BEAM-4071] Add Portable Runner Job API shim

### PR DESCRIPTION
**Pull request by bsidhom**
This runner allows clients of the Java SDK to run against portable an external Job Service with a provided endpoint.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand:
   - [x] What the pull request does
   - [x] Why it does it
   - [x] How it does it
   - [x] Why this approach
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * An {@link AutoCloseable} that wraps a resource that needs to be cleaned up but does not implement
+ * {@link AutoCloseable} itself. Recipients of a {@link CloseableResource} are in general
+ * responsible for cleanup. Not thread-safe.
+ */
+public class CloseableResource<T> implements AutoCloseable {
```
Comment (angoenka) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java]: Should we guard against creation of multiple CloseableResource for same resource?

File: runners/reference/java/src/test/java/org/apache/beam/runners/reference/CloseableResourceTest.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    CloseException closeException = null;
+    try (CloseableResource<Foo> ignored =
+        CloseableResource.of(
+            new Foo(),
+            (foo) -> {
+              throw wrapped;
+            })) {
+      // Do nothing.
+    } catch (CloseException e) {
+      closeException = e;
```
[the remark this answers is not in the document]
Comment (bsidhom) [reply runners/reference/java/src/test/java/org/apache/beam/runners/reference/CloseableResourceTest.java]: Yes, I would much prefer assertThrows. I dislike `ExpectedException` because it limits the assertions that can be made and requires the exception to be thrown as the very last statement of the method. Otherwise, you do not know from which line the exception is actually being thrown.

File: runners/reference/java/src/test/java/org/apache/beam/runners/reference/PortableRunnerTest.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  private static final ApiServiceDescriptor ENDPOINT_DESCRIPTOR =
+      ApiServiceDescriptor.newBuilder().setUrl(ENDPOINT_URL).build();
+
+  private PipelineOptions options = createPipelineOptions();
+
+  @Rule public transient TestPipeline p = TestPipeline.fromOptions(options);
+
+  @Test
+  public void stagesAndRunsJob() throws Exception {
+    try (CloseableResource<Server> server = createJobServer(JobState.Enum.DONE)) {
```
[the remark this answers is not in the document]
Comment (bsidhom) [reply runners/reference/java/src/test/java/org/apache/beam/runners/reference/PortableRunnerTest.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PortablePipelineOptions.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.List;
+import org.apache.beam.sdk.options.Validation.Required;
+
+/** Pipeline options common to all portable runners. */
+public interface PortablePipelineOptions extends PipelineOptions {
+
+  /**
+   * List of local files to make available to workers.
+   *
+   * <p>Files are placed on the worker's classpath.
```
[the remark this answers is not in the document]
Comment (bsidhom) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/PortablePipelineOptions.java]: The worker will access all of the files though the manifest. The files are expected to be jars or directories that should be placed on the classpath and nothing else. This matches the current Dataflow behavior and is the only thing supported.

To really drive this home, we could remove the option to override stage files and just always scan for URLClassLoaders and add elements to the stage file list. However, this prevents users from using custom ClassLoaders. Perhaps the way to solve this for now is to rename this to something like `getClasspathArtifacts` or similar.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PortablePipelineOptions.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+   *
+   * <p>The default value is the list of jars from the main program's classpath.
+   */
+  @Description(
+      "Files to stage to the artifact service and make available to workers. Files are placed on "
+          + "the worker's classpath. The default value is all files from the classpath.")
+  List<String> getFilesToStage();
+  void setFilesToStage(List<String> value);
+
+  @Description("Job service endpoint to use.")
```
[the remark this answers is not in the document]
Comment (bsidhom) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/PortablePipelineOptions.java]: Done.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java [hunk 6]
```diff
@@ -222,13 +225,26 @@ public StagingResult call() {
...
+  @AutoValue
+  public abstract static class FileToStage {
+    public static FileToStage of(File file, String stageName) {
+      return new AutoValue_ArtifactServiceStager_FileToStage(file, stageName);
+    }
+
+    /** The file to stage. */
+    public abstract File getFile();
+    /** Staging handle to this file. */
+    public abstract String getStageName();
```
[the remark this answers is not in the document]
Comment (bsidhom) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java]: No, this is just an opaque handle that the client provides to the staging service. The artifact can be referenced by this name in the manifest.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    checkState(!isClosed, "% is closed", CloseableResource.class.getName());
+    return resource;
+  }
+
+  /**
+   * Returns a new {@link CloseableResource} that owns the underlying resource and relinquishes
+   * ownership from this {@link CloseableResource}. {@link #close()} on the original instance
+   * becomes a no-op.
+   */
+  public CloseableResource<T> transfer() {
```
Comment (bsidhom) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java]: It would be really nice to have the compiler enforce that existing `CloseableResource` references are never passed as arguments but that they are instead always `transfer()`'d in. I'm envisioning something like safe RAII move semantics in C++. Is there a FindBugs check that could enforce this?

Comment (jkff) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java]: > **bsidhom** wrote:
> It would be really nice to have the compiler enforce that existing `CloseableResource` references are never passed as arguments but that they are instead always `transfer()`'d in. I'm envisioning something like safe RAII move semantics in C++. Is there a FindBugs check that could enforce this?


There's no findbugs check that can do that AFAIK. It should be possible to write a check using a compiler plugin, but I think practically speaking we won't :-/

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+/**
+ * An {@link AutoCloseable} that wraps a resource that needs to be cleaned up but does not implement
+ * {@link AutoCloseable} itself.
+ *
+ * <p>Recipients of a {@link CloseableResource} are in general responsible for cleanup. Ownership
+ * can be transferred from one context to another via {@link #transfer()}. Transferring relinquishes
+ * ownership from the original resource. This allows resources to be safely constructed and
+ * transferred within a try-with-resources block. For example:
+ *
+ * <pre>try (CloseableResource&lt;Foo&gt; resource = CloseableResource.of(...)) {
```
Comment (jkff) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java]: <pre>{@code
...
}</pre>
could help you avoid the ampersands.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+ *   // Do something with resource.
+ *   ...
+ *   // Then transfer ownership to some consumer.
+ *   resourceConsumer(resource.release());
```
Comment (jkff) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java]: Did you mean .transfer()?

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InMemoryArtifactService.java [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+  private final Map<ArtifactMetadata, ByteString> artifacts = Maps.newHashMap();
+
+  private final boolean keepArtifacts;
+
+  @GuardedBy("artifactLock")
+  private boolean committed = false;
+
+  /**
+   * Constructs an {@link InMemoryArtifactService}. If {@code keepArtifacts} is true, all artifacts
+   * are kept in memory as {@link ByteString ByteStrings}. Doing so is currently a waste of space
```
Comment (jkff) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InMemoryArtifactService.java]: Hm then can you just remove this parameter? If setting it to true currently accomplishes no useful purpose (it's fine to keep if you plan to very soon follow up this PR with something that makes it have a useful purpose)

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+import org.apache.beam.sdk.runners.PTransformOverride;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** A {@link PipelineRunner} a {@link Pipeline} against a {@code JobService}. */
+public class PortableRunner extends PipelineRunner<PipelineResult> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PortableRunner.class);
+
+  /** Provided pipeline options. */
```
Comment (jkff) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java]: > **bsidhom** wrote:
> I would have gladly done that, but that doesn't fit with the `PipelineRunner` construction. As mentioned above, the fit between a Java runner and a portable runner service isn't great.


Hm okay. I looked at existing PipelineRunner implementations and was unable to figure out whether they do or don't intend to support running multiple pipelines with the same PipelineRunner. We can deal with this later if necessary.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+
+  @Override
+  public String toString() {
+    return "PortableRunner#" + hashCode();
+  }
+
+  private static ManagedChannelFactory getChannelFactory(PipelineOptions options) {
+    ManagedChannelFactory channelFactory;
+    List<String> experiments = options.as(ExperimentalOptions.class).getExperiments();
+    if (experiments != null && experiments.contains("beam_fn_api_epoll")) {
```
Comment (jkff) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java]: > **bsidhom** wrote:
> Not sure. I assume this was copied from https://github.com/apache/beam/blob/15da6412fef89dc08de59828a5df9ba11e944a79/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java#L100
> 
> If we don't have a reason to keep this, we can always use the default factory.


I'd suggest to remove this. This seems like a really old and forgotten experiment, perhaps something Luke was using once for performance testing, and I don't think anybody ever turns it on.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java (hunk 9 above)
Comment (bsidhom) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java]: > **jkff** wrote:
> Did you mean .transfer()?


Done.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java (hunk 8 above)
Comment (bsidhom) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java]: > **jkff** wrote:
> <pre>{@code
> ...
> }</pre>
> could help you avoid the ampersands.


I didn't realize `@code` worked with multiline text. Done.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InMemoryArtifactService.java (hunk 10 above)
Comment (bsidhom) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InMemoryArtifactService.java]: > **jkff** wrote:
> Hm then can you just remove this parameter? If setting it to true currently accomplishes no useful purpose (it's fine to keep if you plan to very soon follow up this PR with something that makes it have a useful purpose)


I've thought about it a bit and realized that we do not want to actually serve artifacts here. That functionality is better handled by a ValidatesRunner integration test. Removed.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java (hunk 12 above)
Comment (bsidhom) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java]: > **jkff** wrote:
> I'd suggest to remove this. This seems like a really old and forgotten experiment, perhaps something Luke was using once for performance testing, and I don't think anybody ever turns it on.


Done.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java (hunk 11 above)
Comment (bsidhom) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java]: > **jkff** wrote:
> Hm okay. I looked at existing PipelineRunner implementations and was unable to figure out whether they do or don't intend to support running multiple pipelines with the same PipelineRunner. We can deal with this later if necessary.


I verified with @tgroh that they do not. Unfortunately, `PipelineOptions` are effectively global flags, so users can't meaningfully instantiate different options anyway.

**Comment by bsidhom** [comment]
Cleaned up commit history. @tgroh PTAL

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java [hunk 13]
```diff
@@ -87,26 +87,32 @@ private ArtifactServiceStager(Channel channel, int bufferSize) {
...
     try {
       StagingResult stagingResult = MoreFutures.get(stagingFuture);
       if (stagingResult.isSuccess()) {
         Manifest manifest =
             Manifest.newBuilder().addAllArtifact(stagingResult.getMetadata()).build();
-        blockingStub.commitManifest(
-            CommitManifestRequest.newBuilder().setManifest(manifest).build());
+        return blockingStub
+            .commitManifest(CommitManifestRequest.newBuilder().setManifest(manifest).build())
+            .getStagingToken();
```
Comment (tgroh) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java]: Separate these two lines? Just a preference thing, but I think that the fluent style hints at something more immediate than the actual RPC we're making here.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java [hunk 14]
```diff
@@ -222,13 +230,26 @@ public StagingResult call() {
     }
   }
 
+  /** A file along with a staging name. */
+  @AutoValue
+  public abstract static class StagedFile {
```
Comment (tgroh) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java]: We create this before staging, right? The name implies after

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (tgroh) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java]: "StagingName"?

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+   */
+  @Override
+  public void close() throws CloseException {
+    if (closer != null && !isClosed) {
+      try {
+        closer.close(resource);
+        isClosed = true;
+      } catch (Exception e) {
+        // Mark resource as closed even if we catch an exception.
+        isClosed = true;
```
Comment (tgroh) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java]: `isClosed` can be in the finally block

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/JobServicePipelineResult.java [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+
+  @Override
+  public State waitUntilFinish(Duration duration) {
+    if (duration.compareTo(Duration.millis(1)) < 1) {
+      // Equivalent to infinite timeout.
+      return waitUntilFinish();
+    } else {
+      CompletableFuture<State> result = CompletableFuture.supplyAsync(this::waitUntilFinish);
+      try {
+        return Uninterruptibles.getUninterruptibly(
```
Comment (tgroh) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/JobServicePipelineResult.java]: Why is this uninterruptible?

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java (hunk 11 above)
Comment (tgroh) [reply runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java]: They do, but always with the same options.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+  private final ManagedChannelFactory channelFactory;
+
+  /**
+   * Constructs a runner from the provided options.
+   *
+   * @param options Properties which configure the runner.
+   * @return The newly created runner.
+   */
+  public static PortableRunner fromOptions(PipelineOptions options) {
+    return createInternal(options, ManagedChannelFactory.createDefault());
```
Comment (tgroh) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java]: why not just `create`?

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+    // HACK: Encode the path name ourselves because the local artifact staging service currently
+    // assumes artifact names correspond to a flat directory. Artifact staging services should
+    // generally accept arbitrary artifact names.
+    // NOTE: Base64 url encoding does not work here because the stage artifact names tend to be long
+    // and exceed file length limits on the artifact stager.
+    String encodedPath = escapePath(file.getPath());
+    return StagedFile.of(file, encodedPath);
+  }
+
+  /** Create a filename-friendly artifact name for the given path. */
```
Comment (tgroh) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java]: It really seems like there's gotta be either a library way to do this, or some clever workaround?

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InMemoryArtifactService.java [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+import org.apache.beam.model.jobmanagement.v1.ArtifactApi.ArtifactMetadata;
+import org.apache.beam.model.jobmanagement.v1.ArtifactApi.CommitManifestRequest;
+import org.apache.beam.model.jobmanagement.v1.ArtifactApi.CommitManifestResponse;
+import org.apache.beam.model.jobmanagement.v1.ArtifactApi.PutArtifactRequest;
+import org.apache.beam.model.jobmanagement.v1.ArtifactApi.PutArtifactResponse;
+import org.apache.beam.model.jobmanagement.v1.ArtifactStagingServiceGrpc.ArtifactStagingServiceImplBase;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** A StagingService for tests. Only stores artifact metadata. */
```
Comment (tgroh) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InMemoryArtifactService.java]: We already have one of these, don't we?

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java (hunk 14 above)
Comment (bsidhom) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java]: > **tgroh** wrote:
> We create this before staging, right? The name implies after


See https://github.com/apache/beam/pull/5150#discussion_r182507804. I'm open to suggestions, but otherwise I'll leave it as is.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java (hunk 18 above)
Comment (bsidhom) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java]: > **tgroh** wrote:
> It really seems like there's gotta be either a library way to do this, or some clever workaround?


This _is_ the clever workaround.

I actually tried base64 encoding, but it turns out that very long path names end up on the classpath and base64 encoding causes path names to exceed size limits on the systems I tried it on. See https://github.com/apache/beam/commit/13ce9cf872e2856fbddff6822b5017a53bbd90f6.

The proper fix is to support arbitrary staging names server-side by maintaining an artifact name mapping independent of filesystem rules and limits. I created BEAM-4109 to track this; it should really be its own change.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java (hunk 13 above)
Comment (bsidhom) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java]: > **tgroh** wrote:
> Separate these two lines? Just a preference thing, but I think that the fluent style hints at something more immediate than the actual RPC we're making here.


Done.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/JobServicePipelineResult.java (hunk 16 above)
Comment (bsidhom) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/JobServicePipelineResult.java]: > **tgroh** wrote:
> Why is this uninterruptible?


To ensure that we sleep fully for the amount of time requested. Given that we don't use interrupts anywhere in Beam code, we don't expect this to happen anyway. Alternatively we could just crash on interrupts (i.e., throw a RuntimeException wrapping the interruption).

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InMemoryArtifactService.java (hunk 19 above)
Comment (bsidhom) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InMemoryArtifactService.java]: > **tgroh** wrote:
> We already have one of these, don't we?


I was unable to find one. Please link it if it does exist.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java (hunk 15 above)
Comment (bsidhom) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java]: > **tgroh** wrote:
> `isClosed` can be in the finally block


Done.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java (hunk 17 above)
Comment (bsidhom) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java]: > **tgroh** wrote:
> why not just `create`?


Mostly to draw attention to the fact that this is an internal call and to distinguish it from `fromOptions`. If you have a strong preference, I'll change it.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java (hunk 14 above)
Comment (bsidhom) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java]: > **tgroh** wrote:
> We create this before staging, right? The name implies after


See https://github.com/apache/beam/pull/5150#discussion_r182507804. I'm open to suggestions, but otherwise I'll leave it as is.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java (hunk 17 above)
Comment (bsidhom) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java]: > **tgroh** wrote:
> why not just `create`?


Mostly to draw attention to the fact that this is an internal call and to distinguish it from `fromOptions`. If you have a strong preference, I'll change it.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java (hunk 18 above)
Comment (bsidhom) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java]: > **tgroh** wrote:
> It really seems like there's gotta be either a library way to do this, or some clever workaround?


This _is_ the clever workaround.

I actually tried base64 encoding, but it turns out that very long path names end up on the classpath and base64 encoding causes path names to exceed size limits on the systems I tried it on. See https://github.com/apache/beam/commit/13ce9cf872e2856fbddff6822b5017a53bbd90f6.

The proper fix is to support arbitrary staging names server-side by maintaining an artifact name mapping independent of filesystem rules and limits. I created BEAM-4109 to track this; it should really be its own change.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java (hunk 13 above)
Comment (bsidhom) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java]: > **tgroh** wrote:
> Separate these two lines? Just a preference thing, but I think that the fluent style hints at something more immediate than the actual RPC we're making here.


Done.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/JobServicePipelineResult.java (hunk 16 above)
Comment (bsidhom) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/JobServicePipelineResult.java]: > **tgroh** wrote:
> Why is this uninterruptible?


To ensure that we sleep fully for the amount of time requested. Given that we don't use interrupts anywhere in Beam code, we don't expect this to happen anyway. Alternatively we could just crash on interrupts (i.e., throw a RuntimeException wrapping the interruption).

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java (hunk 15 above)
Comment (bsidhom) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/CloseableResource.java]: > **tgroh** wrote:
> `isClosed` can be in the finally block


Done.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InMemoryArtifactService.java (hunk 19 above)
Comment (bsidhom) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InMemoryArtifactService.java]: > **tgroh** wrote:
> We already have one of these, don't we?


I was unable to find one. Please link it if it does exist.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/JobServicePipelineResult.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (tgroh) [reply runners/reference/java/src/main/java/org/apache/beam/runners/reference/JobServicePipelineResult.java]: I strongly favor the latter - `InterruptedException` generally means you're being asked to shut down, which we should respect unless we have *extremely strong reasons* to not; this plays nicely with things like `ExecutorService.shutdown()`, and related.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/JobServicePipelineResult.java [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+  @Override
+  public State waitUntilFinish() {
+    JobServiceBlockingStub stub = jobService.get();
+    GetJobStateRequest request = GetJobStateRequest.newBuilder().setJobIdBytes(jobId).build();
+    GetJobStateResponse response = stub.getState(request);
+    State lastState = getJavaState(response.getState());
+    while (!lastState.isTerminal()) {
+      Uninterruptibles.sleepUninterruptibly(POLL_INTERVAL_SEC, TimeUnit.SECONDS);
```
Comment (tgroh) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/JobServicePipelineResult.java]: Just normal sleeps, breaking out of the method if we're interrupted.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java (hunk 17 above)
Comment (tgroh) [reply runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java]: It's implied to be internal because it's package private and marked `@VisibleForTesting` - anyone using it outside of developers in this package is already jumping into a volcano in terms of behaving properly

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (tgroh) [reply runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java]: Please link that issue in the TODO

[the remark this answers is not in the document]
Comment (tgroh) [reply runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java]: And I'd be kind of sad, but differently (and maybe a bit less), by just replacing names that are too bad with like, b64-encoded hashes or something (though also that would be _extremely confusing_ to debug, so probably more sad)

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InMemoryArtifactService.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (tgroh) [reply runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InMemoryArtifactService.java]: https://github.com/apache/beam/blob/master/runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/InMemoryArtifactStagerService.java

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InProcessManagedChannelFactory.java [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ * A {@link org.apache.beam.sdk.fn.channel.ManagedChannelFactory} that uses in-process channels.
+ *
+ * <p>The channel builder uses {@link ApiServiceDescriptor#getUrl()} as the unique in-process name.
+ */
+public class InProcessManagedChannelFactory extends ManagedChannelFactory {
+
+  @Override
+  public ManagedChannel forDescriptor(ApiServiceDescriptor apiServiceDescriptor) {
```
Comment (tgroh) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InProcessManagedChannelFactory.java]: I've been using a lambda when this is called for (e.g. in https://github.com/apache/beam/blob/master/runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/RemoteExecutionTest.java#L124); mind updateing those also?

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/TestJobService.java [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+  @Override
+  public void getState(
+      GetJobStateRequest request, StreamObserver<GetJobStateResponse> responseObserver) {
+    responseObserver.onNext(GetJobStateResponse.newBuilder().setState(jobState).build());
+    responseObserver.onCompleted();
+  }
+
+  @Override
+  public void cancel(CancelJobRequest request, StreamObserver<CancelJobResponse> responseObserver) {
+    super.cancel(request, responseObserver);
```
Comment (tgroh) [inline runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/TestJobService.java]: Remove overrides that are exactly equal to the `super` method

File: runners/reference/java/src/test/java/org/apache/beam/runners/reference/CloseableResourceTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (tgroh) [reply runners/reference/java/src/test/java/org/apache/beam/runners/reference/CloseableResourceTest.java]: All of your tests here match exactly that pattern, though - so I'm not super convinced that we want to have the one test which isn't like the other ones?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PortablePipelineOptions.java [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+package org.apache.beam.sdk.options;
+
+import java.util.List;
+import org.apache.beam.sdk.options.Validation.Required;
+
+/** Pipeline options common to all portable runners. */
+public interface PortablePipelineOptions extends PipelineOptions {
+
+  // TODO: https://issues.apache.org/jira/browse/BEAM-4106: Consider pulling this out into a new
+  // options interface, e.g., FileStagingOptions.
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/PortablePipelineOptions.java]: Highly likely that we also desire autopopulation of the classpath here, so worth a comment there also

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InMemoryArtifactService.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (bsidhom) [reply runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InMemoryArtifactService.java]: Wow. I guess IntelliJ failed me. 😢

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InProcessManagedChannelFactory.java (hunk 21 above)
Comment (bsidhom) [reply runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/InProcessManagedChannelFactory.java]: Done.

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/TestJobService.java (hunk 22 above)
Comment (bsidhom) [reply runners/reference/java/src/main/java/org/apache/beam/runners/reference/testing/TestJobService.java]: Done.

File: runners/reference/java/src/test/java/org/apache/beam/runners/reference/CloseableResourceTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bsidhom) [reply runners/reference/java/src/test/java/org/apache/beam/runners/reference/CloseableResourceTest.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PortablePipelineOptions.java (hunk 23 above)
Comment (bsidhom) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/PortablePipelineOptions.java]: Done.

**Comment by bsidhom** [comment]
retest this please

**Comment by bsidhom** [comment]
Commits should be clean now.

**Comment by tgroh** [comment]
Gotta rebase one more time, I fear

**Comment by bsidhom** [comment]
Done.

File: runners/reference/java/build.gradle [hunk 24]
```diff
@@ -27,8 +27,11 @@ framework to execute user-definied functions."""
 dependencies {
   shadow project(path: ":beam-model-pipeline", configuration: "shadow")
   shadow project(path: ":beam-runners-core-construction-java", configuration: "shadow")
+  shadow project(path: ":beam-sdks-java-fn-execution", configuration: "shadow")
   shadow library.java.slf4j_api
-  testCompile library.java.junit
+  shadowTest project(path: ":beam-runners-core-construction-java", configuration: "shadowTest")
```
Comment (kennknowles) [inline runners/reference/java/build.gradle]: Why is this here? We should not have these kinds of dependencies except for the purpose of running test suites.

Comment (kennknowles) [reply runners/reference/java/build.gradle]: (is it pulling in NeedsRunner tests? if so, comment that, since these dependencies are by default an incorrect thing to do)

Comment (bsidhom) [reply runners/reference/java/build.gradle]: No. This is pulling in `InMemoryArtifactStagerService`, which is a test _utility_ rather than a test itself. That class should really live under the main source set in a `testing` subpackage, but I didn't want to do that refactor here.

Is there a tracking bug for removing cross-module test dependencies?

Comment (kennknowles) [reply runners/reference/java/build.gradle]: Yes - https://issues.apache.org/jira/browse/BEAM-3138

Comment (kennknowles) [reply runners/reference/java/build.gradle]: For now, IMO it is fine to just move it to `src/main/java` and mark `@Internal` if you don't want to do any other edits.
